src/cowboy_req.erl

```diff
@@ -666,6 +671,126 @@ delete_resp_header(Name, Req=#http_req{resp_headers=RespHeaders}) ->
+	catch throw:_Reason ->
+		Req
+	end.
+
+-spec set_cors_allow_credentials(boolean(), {binary(), binary(), 0..65535} | reference(), Req) -> Req when Req :: req(). 
+set_cors_allow_credentials(Credentials, Origin, Req) ->
+	case match_cors_credentials(Credentials, Origin) of
+		true ->
+			Req2 = set_resp_header(<<"access-control-allow-origin">>, cow_http_hd:access_control_allow_origin(Origin), Req),
+			set_resp_header(<<"access-control-allow-credentials">>, cow_http_hd:access_control_allow_credentials(), Req2);
+		_ ->
+			set_resp_header(<<"access-control-allow-origin">>, cow_http_hd:access_control_allow_origin(Origin), Req)
+	end.
+
+-spec set_cors_max_age(non_neg_integer() | undefined, Req) -> Req when Req :: req().
+set_cors_max_age(undefined, Req) ->
+	Req;
+set_cors_max_age(Val, Req) ->
+	set_resp_header(<<"access-control-max-age">>, cow_http_hd:access_control_max_age(Val), Req).
+
+-spec set_cors_allowed_methods([binary()], Req) -> Req when Req :: req().
+set_cors_allowed_methods(L, Req) ->
+	set_resp_header(<<"access-control-allow-methods">>, cow_http_hd:access_control_allow_methods(L), Req).
+
+-spec set_cors_allowed_headers([binary()], Req) -> Req when Req :: req().
+set_cors_allowed_headers([], Req) ->
+	Req;
+set_cors_allowed_headers(L, Req) ->
+	set_resp_header(<<"access-control-allow-headers">>, cow_http_hd:access_control_allow_headers(L), Req).
+
+-spec set_cors_exposed_headers([binary()], Req) -> Req when Req :: req().
+set_cors_exposed_headers([], Req) ->
+	Req;
+set_cors_exposed_headers(L, Req) ->
+	set_resp_header(<<"access-control-expose-headers">>, cow_http_hd:access_control_expose_headers(L), Req).
+
+-spec match_cors_origin(Origin | reference(), [Origin] | Origin | '*')
+	-> Origin | '*' when Origin :: {binary(), binary(), 0..65535}.
+match_cors_origin(Val, '*') when is_reference(Val) ->
+	'*';
+match_cors_origin(Val, '*') ->
+	Val;
+match_cors_origin(Val, Val) ->
+	Val;
+match_cors_origin(Val, AllowedOrigins) when is_list(AllowedOrigins) ->
+	case lists:member(Val, AllowedOrigins) of
+		true -> Val;
+		_ -> throw({nomatch_origin, Val, AllowedOrigins})
+	end;
+match_cors_origin(Val, AllowedOrigins) ->
+	throw({nomatch_origin, Val, AllowedOrigins}).
+
+-spec match_cors_method(binary() | undefined, [binary()]) -> binary().
+match_cors_method(undefined, Methods) ->
+	throw({bad_method, undefined, Methods});
+match_cors_method(Val, AllowedMethods) ->
+	case lists:member(Val, AllowedMethods) of
+		true -> Val;
+		_ -> throw({nomatch_method, Val, AllowedMethods})
+	end.
+
+-spec match_cors_headers([binary()], [binary()]) -> [binary()].
+match_cors_headers(L, AllowedHeaders) ->
+	[case lists:member(Header, AllowedHeaders) of
+		false -> throw({nomatch_header, Header, AllowedHeaders});
+		_ -> Header
+	end || Header <- L].
+
+-spec match_cors_credentials(boolean(), {binary(), binary(), 0..65535} | reference() | '*') -> boolean().
+match_cors_credentials(true, '*') ->
+	throw({bad_credentials, true, '*'});
```

Comment:
Guessing this is for 'The string "*" cannot be used for a resource that supports credentials.' but I'm not sure this is very useful, more useful would be sending Origin back instead of "*" if credentials are allowed.